pkg/template/controller/templateinstance_controller.go

```diff
@@ -1,124 +1,208 @@
+		lister:         informer.Lister(),
+		informer:       informer.Informer(),
+		queue:          workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "TemplateInstanceController"),
+	}
+
+	c.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			c.enqueue(obj.(*templateapi.TemplateInstance))
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			c.enqueue(obj.(*templateapi.TemplateInstance))
 		},
-		&templateapi.TemplateInstance{},
-		0,
-		cache.ResourceEventHandlerFuncs{
-			AddFunc: func(obj interface{}) {
-				c.handle(obj.(*templateapi.TemplateInstance))
-			},
-			UpdateFunc: func(_, obj interface{}) {
-				c.handle(obj.(*templateapi.TemplateInstance))
-			},
-			DeleteFunc: func(obj interface{}) {
-			},
+		DeleteFunc: func(obj interface{}) {
 		},
-	)
+	})
 
-	return &c
+	return c
 }
 
-func (c *TemplateInstanceController) Run(stop <-chan struct{}) {
-	c.controller.Run(stop)
+func (c *TemplateInstanceController) getTemplateInstance(key string) (*templateapi.TemplateInstance, error) {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.lister.TemplateInstances(namespace).Get(name)
 }
 
-func (c *TemplateInstanceController) handle(templateInstance *templateapi.TemplateInstance) error {
+func (c *TemplateInstanceController) copyTemplateInstance(templateInstance *templateapi.TemplateInstance) (*templateapi.TemplateInstance, error) {
+	templateInstanceCopy, err := kapi.Scheme.DeepCopy(templateInstance)
+	if err != nil {
+		return nil, err
+	}
+
+	return templateInstanceCopy.(*templateapi.TemplateInstance), nil
+}
+
+func (c *TemplateInstanceController) sync(key string) error {
+	templateInstance, err := c.getTemplateInstance(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
 	for _, condition := range templateInstance.Status.Conditions {
 		if condition.Type == templateapi.TemplateInstanceReady && condition.Status == kapi.ConditionTrue ||
 			condition.Type == templateapi.TemplateInstanceInstantiateFailure && condition.Status == kapi.ConditionTrue {
 			return nil
 		}
 	}
 
-	err := c.provision(templateInstance)
-	if err == nil {
-		templateInstance.Status.Conditions = []templateapi.TemplateInstanceCondition{
-			{
-				Type:               templateapi.TemplateInstanceReady,
-				Status:             kapi.ConditionTrue,
-				LastTransitionTime: metav1.Now(),
-				Reason:             "Created",
-			},
-		}
+	templateInstance, err = c.copyTemplateInstance(templateInstance)
+	if err != nil {
+		return err
+	}
+
+	provisionErr := c.provision(templateInstance)
+
+	templateInstance.Status.Conditions = templateapi.FilterTemplateInstanceCondition(templateInstance.Status.Conditions, templateapi.TemplateInstanceReady)
+	templateInstance.Status.Conditions = templateapi.FilterTemplateInstanceCondition(templateInstance.Status.Conditions, templateapi.TemplateInstanceInstantiateFailure)
```

Comment:
yeah i was wondering if they were, but i agree w/ the goal of certainty, so i think it's fine to leave them.